### ISSUE TITLE
[Bug] vimium's `window.forTrusted()` blocks javascript-dispatched events **not only from itself, but from the whole page**

### ISSUE THREAD

**Issue by kiprasmel** [philc/vimium#3651]
**To Reproduce**

Example in chrome; also reproduces the same way in firefox.

1. Go to a site where vimium works (i.e. [example.com](http://example.com))

2. Devtools > Sources > Enable Listener Breakpoints > Keyboard (check)

3. Console & dispatch a sample event:

```js
event = new KeyboardEvent("keydown", { key: "Enter" });

element = document.querySelector("body");

element.dispatchEvent(event);
```

1. Get thrown back into Sources, the `utils.js` file from vimium (full path `chrome-extension://dbepggeogbaibhgnhhndojpepiihcmeb/lib/utils.js`)

2. See the event get ignored instead of getting handled, since `event.isTrusted !== true`.

3. Explore further with F10; exit from breakpoint with F8

4. See `true` get returned from the same place.

That's essentially it. I myself encountered this at gitlab, and, well, I ain't getting those 7 hours of debugging back, so I hope we can fix this so that others don't need to:D

**Browser and Vimium version**

```sh
$ /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --version
Google Chrome 84.0.4147.135

$ /Applications/Firefox\ Developer\ Edition.app/Contents/MacOS/firefox  --version
Mozilla Firefox 81.0b2
```

vimium version 1.66 in both

**Comment by gdh1995** [comment]
Why do you think it's a bug? It seems safe and very useful if Vimium ignores untrusted events from any source.

**Comment by kiprasmel** [comment]
Vimium, for itself, can do whatever it wants. The problem I have is that ALL events get blocked, even if I'm intentionally using them.

**Comment by gdh1995** [comment]
do you mean event.stopProp or preventDefault?&nbsp;
As far as I can remember,&nbsp; Vimium just ignores such events and lets them continue to pass to other listeners.&nbsp; Is there anything wrong?



---Original---

**Comment by kiprasmel** [comment]
I mean the exact reproduction case that I provided - I'm not sure if's got anything to do with `event.stopProp`, but definitely not `event.preventDefault` -- the main problem is that vimium, in the reproduction case above, does **not** let the events to pass through.

**Comment by gdh1995** [comment]
well, this is by intent, because web pages are in fact not so reliable, and those events of .isTrusted=false may come from untrusted JS code, like an AD card's.


there's no way to tell from a event created by AD cards from yours created on Console.



---Original---

**Comment by kiprasmel** [comment]
Yeah well I agree the security part is good, but I'm not willing to give up usability just for the sake of it, and would ideally prefer to have a toggle for the !isTrusted event (un)blocking.

**Comment by lydell** [comment]
I’m curious – what’s the use case for having websites be able to trigger Vimium commands?

**Comment by kiprasmel** [comment]
The use case is **not** for websites to trigger **vimium** commands - it's to trigger **any** events. **vimium blocks all events that are not .isTrusted!**

Have you all tried the reproduction case? It seems like you're missing the point?

**Comment by lydell** [comment]
Yeah, I have read it but it’s difficult to understand. Could you make a demo page, maybe? (Note: I’m not a maintainer.)

**Comment by gdh1995** [comment]
I know your case is reproduceable,&nbsp; Um, but what's your real use case?&nbsp; Do you want an untrusted mouse event to let Vimium exit the GrabBackFocus mode, or anything else?



---Original---

**Comment by kiprasmel** [comment]
> Yeah, I have read it but it’s difficult to understand. Could you make a demo page, maybe? (Note: I’m not a maintainer.)

yes it is difficult to understand it took me 8 hours to find it



> I know your case is reproduceable,&nbsp; Um, but what's your real use case?&nbsp; Do you want an untrusted mouse event to let Vimium exit the GrabBackFocus mode, or anything else?

My use case is this: I am creating my own browser extension for GitLab, and it's API lacks sometimes, so I'm using the user interface directly with events. If I have the Vimium extension enabled, it blocks those events and my extension loses some functionality, and I don't want that, obviously.

**Comment by gdh1995** [comment]
So you were using your extension to click a button on GitLab, but Vimium blocks such keyboard events so the clicking didn't work? And, when Vimium is disabled, do the simulated events work as expected?

**Comment by kiprasmel** [comment]
@gdh1995 yes - exactly.

**Comment by gdh1995** [comment]
It's very strange because Vimium just 'ignore' untrusted events, that is, such events should be passed to the page normally.


&nbsp;Could you give a demo in which a part of js code fails to do something, if only Vimium is enabled?



---Original---

**Comment by afilp** [comment]
@gdh1995 
I actually have a use case where I need to accress the vimium API programmatically.

I want my test page to open <c-m> automatically, on load and on getting the focus, so that I can immediately search for any open web page in my tabs.

See a test page here, with explanation of the goal here:
https://www.likno.com/test/blank/

Maybe you have a different way to access the "vimium API" programmatically, than trying to simulate a keypress. I understand the limitations of the extensions sandbox though. What if we could "white-label" a specific page (or pattern) to allow the keyboard events even if they are not trusted?

a wanted the "b" to be pressed automatically when I

**Comment by kiprasmel** [comment]
@afilp create a separate issue -- this one not yet resolved.

**Comment by afilp** [comment]
Done: https://github.com/philc/vimium/issues/3710  Thanks.
